std/internal/math/gammafunction.d

```diff
@@ -548,7 +550,11 @@ unittest {
 
 
 private {
-static if (floatTraits!(real).realFormat == RealFormat.ieeeExtended) {
+static if (floatTraits!(real).realFormat == RealFormat.ieeeQuadruple) {
+    enum real MAXLOG = 0x1.62e42fefa39ef35793c7673007e6p+13;  // log(real.max)
+    enum real MINLOG = -0x1.6546282207802c89d24d65e96274p+13; // log(real.min_normal*real.epsilon) = log(smallest denormal)
```

Comment:
@redstar Could you explain please what is this constants and how they was computed?

Reply:
I just reverse engineered this, there are obviously too many things you can do wrong when calculating this value...
* Make sure `log` is the natural logarithm!
* `real.max is really the max value`, but you need the exact value (e.g. in binary representation, not some decimal approximation). For 64 bit: `2^1023·(1 + (1 − 2^−52))`. Note: you can get these from the D compiler: `writefln!"%a"(real.max or real.man_dig or real.epsilon);` the returned value is a bit exact result. (But if there's a calculation involved, the calculation is of course affected by the FPUs range and rounding). Remember to convert for wolfram alpha: `0xf.fffffffffffffffp+16380 ==> (f.fffffffffffffff base 16) * 2^16380`
* The hex format is `X.XXXX * 2^y`, so mantissa X needs to be base 16 and exponent y needs to be base 2! This is not the wolfram alpha base 16 format!
* You cannot simply calculate this with standard math functions in C/D/... because of limited precision!
* Calculate base 2 representation in wolfram alpha, then convert the base 2 mantissa to base 16
64 bit: http://www.wolframalpha.com/input/?i=ln(2%5E1023%C2%B7(1+%2B+(1+%E2%88%92+2%5E%E2%88%9252)))+in+base+2
80 bit: http://www.wolframalpha.com/input/?i=ln((f.fffffffffffffff+base+16)+*+2%5E16380)+in+base+2
128 bit: http://www.wolframalpha.com/input/?i=ln((1.ffffffffffffffffffffffffffff+base+16)+*+2%5E16383)+in+base+2

```d
void main()
{
    import std.algorithm, std.stdio, std.conv;
    string mantissa = "1.01100010111001000010111111101111101000111001111011110011010100111001001111000111011001110011000000000111111001011101110101011";
    write(mantissa[0 .. 2]); // always 1.
    mantissa = mantissa[2 .. $];
    for (size_t i = 0; i < mantissa.length/4; i++)
    {
        writef!"%x"(to!ubyte(mantissa[0..4], 2));
        mantissa = mantissa[4 .. $];
    }
    writeln();
}
```


To summarize: I hate the internal/math modules. It delayed the ARM port years ago, now it delayed the AArch64 port for 2 years. Nobody maintains this code, nobody knows what it's supposed to do, the referenced original C code doesn't match this code, so it isn't useful either.... I doubt this code is even correct. For example, in `gammafunction.d` we have `MAXLOG` for extended/double format. But in `errorfunction.d` we use only the extended `MAXLOG` for all float types. Is this correct? I guess not, I don't know and I don't know who could know. The whole math module situation is a mess. ping @andralex I think we need some long term solution for these modules....

For now, as this is the last thing stopping GDC (and probably LDC) from compiling phobos on AArch64 I'll post a revived version of this pull request tomorrow.